### ISSUE TITLE
Proposal for a common attached-services concept

### ISSUE THREAD

**Issue by miraculixx** [progrium/dokku#854]
I'm not sure proposals like this are warranted, so please let me know if not. This is born out of the observation (and some frustration ;-) that various dokku plugins take different (and somewhat conflicting) approaches in both naming convention, command syntax and functionality. 

### Goal

Facilitate an ecosystem of dokku plugins that provide [attached services](http://12factor.net/backing-services) that follow a common pattern in building such plugins, their use cases and a common command syntax. This is intended as a minimum level.

### Rationale
Currently there seem to be different ways for attached services. Some prefer to attach a service to a single app by default, while the dokku-link plugin allows one service to be linked to multiple apps. Also there are various ways for service plugins to name their environment variables.

This is to propose a standard for attached services, as follows. Ammendments, feedback and proposals welcome.

### Definition

1. A service is a dokku-plugin that, upon plugin installation, builds a base image to run the service in its own container.
2. A service can be linked to any number of dokku apps, but it can also be created without any app specified (deferred linking).
3. When linked, the service is to expose its connection URL to app containers as a single environment variable that follows this dokku-args standard: `-e <SERVICE>_URL=<service>://<user>:<password>@<ip>[/<path>]`, where service is the service name (e.g. MYSQL, RABBITMQ etc.), user/password the access credentials or tokens, ip the linked container's address, and the optional path any further information required.
4. Services can expose further name-spaced env variables via dokku-args or dokku's `config:set` command, e.g. `-e <SERVICE>_SOMEVAR=some-value`. Such variables are to be documented in the plugin's README for easy reference
5. Upon startup of the service container, the plugin creates a credentials file in $DOKKU_ROOT/.$SERVICE/credentials
6. Upon link of a service to an app, the plugin uses dokku-link to create the link
7. Upon removal of a service link from an app, the plugin uses dokku-link to remove the link
8. A service maintains its persistent data (if any) in $DOKKU_ROOT/.$SERVICE/volume, which it maps in its container as /var/$SERVICE. Such volumes are to be recorded in $DOKKU_ROOT/.$SERVICE/volumes
9. A service offers import and export commands to allow users to import/export service state (e.g. backup of persistent data or configuration).
10. A service can be cloned. Cloning means to create a new service that exactly matches the specification of the old service, and copies all data (i.e. does <service>:export <old>, followed by <service>:import <new>>`.

### Common command syntax

`<service>:create <name> [<options>]` => create the service, no link to an app. options are optional and service-specific
`<service>:link <name> <app>` => link the service to an app
`<service>:unlink <name> <app>` => unlink the service from an app
`<service>:delete> <name>` => if there are no links left, delete the service and stop its container
`<service>:export <name>` => return a the export filename that can be used by import  (URL?)
`<service>:import <name> <file>` => import file into the service (deleting all previous data first)
`<service>:logs <name>` => print the most recent log(s) for this service
`<service>:restart <name>` => graceful shutdown and restart of the service container. graceful means it will only work if no other containers/apps are currently using the service.
`<service>:info <name>` => print the connection information
`<service>:list` => list all services of this type
`<service>:clone <existing-name> <new-name> [config | data | both]` => shortcut for :create the service new-name by copying the config (default), the data (do :export <name>, then :import <new-name>), or both from the existing service <name> . The new service shall *not* be linked to any apps. 
`<service>:expose <name> port` => expose the service on port. By default a service does not publicly expose any ports and can only be access via docker links. 

### Benefits
- attached services follow a common pattern that is simple and effective, making it easier for users to build upon the respective plugins.
- makes it easier to implement services as you don't have to re-invent the wheel
- by specifically using the `docker link` facilities services become first class citizen in the docker and dokku universes (like apps).

### Acknowledgment 

This is meant as a constructive first step, and very much in appreciation of all those who have provided service plugins. In particular this is inspired and builds upon the fine work done by the authors of [dokku-redis](https://github.com/luxifer/dokku-redis-plugin), [dokku-mysql](https://github.com/hughfletcher/dokku-mysql-plugin), [dokku-rabbitmq](https://github.com/jlachowski/dokku-rabbitmq-plugin) and [dokku-link](https://github.com/rlaneve/dokku-link).

Updates
* added :clone syntax
* added :expose syntax as [patrickjahn](https://github.com/progrium/dokku/issues/854#issuecomment-94611560) suggested

**Comment by josegonzalez** [comment]
@miraculixx can you take a look at #292 ? Would be nice to combine the two. Also, I'd love to see this as a PR :)

**Comment by mlebkowski** [comment]
Not so long ago I was deploying a large application using dokku. It had a lot of attached services (mysql, mongo, elasticsearch, memcache to name a few). I believe that it should be the applications responsibility to define those services and a dokku plugin to **automatically** launch them during the release.

I’ve done some crude work towards that goal here: https://github.com/mlebkowski/dokku/blob/master/plugins/10_services/pre-release

There are of course specific circumstances:
 * It’s only targeting PHP / composer apps (configuration is stored in `composer.json`), but it could easily be moved elsewhere
 * It requires `jq` to be present in the built image (I think all the plugin’s heavy lifting should be done in dedicated containers anyway)
 * It doesn’t care at all about the volume persistence, since I only needed short-lived test deploys

But besides that I think that the approach is better. Defining those dependencies in application instead of launching the services by hand. [Wercker does a similar thing](http://devcenter.wercker.com/articles/services/) and it works great there. IMMV.

**Comment by miraculixx** [comment]
@mlebkowski thanks for your comment and input.

>  But besides that I think that the approach is better. Defining those dependencies in application instead of launching the services by hand.

agree, but this should be built _on top_ of the services infrastructure, not by replacement. Once there are commands to start/attach services by command it is trivial to build an automated solution. The opposite, not so much.

**Comment by josegonzalez** [comment]
@miraculixx do you have a reference plugin or plugins we can look at? How do you propose services like RDS or Elasticache fit into this? What if I made a service plugin that integrated, say, heroku hosted postgres?

**Comment by miraculixx** [comment]
@josegonzalez I'm afraid I don't have a ready reference yet. Working on that...

**Comment by mlebkowski** [comment]
Just wanted to let you know that wercker is switching to docker for it’s infrastructure. Their config is dead simple:

```
box: ubuntu:latest
services:
  - dockerfile/mysql
  - dockerfile/elasticsearch
```

Maybe they are willing to contribute? /cc @mies.

**Comment by josegonzalez** [comment]
We tried testing on wercker before for dokku, and their testing infra was spotty - failed builds more often than not - and we're using CircleCI now to great success. Not that we wouldn't push our users to use what works best.

**Comment by patrickjahns** [comment]
@mlebkowski 

Any further updates on your idea? I`ve been building some setups with docker these days and something that frustrates me is, that I need to to a lot of manual setup on the server itself to link containers or orchestrate services in the containers.

I`d really like to have a configuration file for the service setup. (Similar to how travis, circle-ci or drone-ci do it). At drone-ci they are working on something called "matrix builds" - I think we can look at the way they read their configuration files and adapt them for dokku? 

I am not sure how the process can be done with bash though...

**Comment by miraculixx** [comment]
>  am not sure how the process can be done with bash though...

Indeed, I'm working on a Python based approach for services as I described above. Not sure if that will be welcomed here, but I found bash starting to get too complicated for what I think it needs to be...

**Comment by progrium** [comment]
First, this is about what I'd call "addons" as that is the Heroku equivalent. However, I think that should be different than support for docker-compose/fig, which this is turning into.

**Comment by patrickjahns** [comment]
@miraculixx 
As soon as you have something, please share. I was also thinking of starting a couple of things from scratch - but in the end dokku does something good, with a couple of hickups here and there, but I am very surprised about how far bash scripting can actually go.

Back on topic: 
What I love from the drone-ci syntax is, that you can specify direct commands to be run in the container when provisioned. These commands can be anything that your shell can do and provides a lot of flexibility that way 

@progrium 
How do you see docker compose/fig playing together with dokku? 
Will we be easily able to for example creating an "application" specify in the fig file that we want service x, service y, mount volume f, open port z ? Or will it merely allow us to link containers

I am really keen on handling the configuration via the sourcecode of the application. For me living inside the code feels more natural

**Comment by patrickjahns** [comment]
@mlebkowski 

I had another thought on your proposal for services, I would like to add also, that each service would show a list to which apps/containers it is linked to.

**Comment by mlebkowski** [comment]
@patrickjahns, nope, I didn’t work on this since.

But maybe instead of reinventing the wheel, we could use [docker compose](https://docs.docker.com/compose/)?

**Comment by matthewmueller** [comment]
@progrium 

> First, this is about what I'd call "addons" as that is the Heroku equivalent. However, I think that should be different than support for docker-compose/fig, which this is turning into.

Can you clarify what you mean here? My take is that services (addons) are more like linking containers while docker-compose/fig is more of a blueprint to launch your multi-container application. Is that what you mean?

**Comment by progrium** [comment]
Yes. In the Docker world an "addon" might be just another container service in docker-compose, but in Heroku, it's implicit and defined via environment and could even be remote. So there should be two features: docker-compose support, and ideally Heroku-style addons. However, I think docker-compose support would satisfy demand for addons since it will solve the problem for most people.

So I'm saying I'd like to see either a more fully implemented Heroku-style addon plugin or just plain docker-compose support. And not call either one "attached-services".

**Comment by josegonzalez** [comment]
So basically https://github.com/project-nsmg/figure ?

**Comment by progrium** [comment]
Yeah. Seems like that should be a plugin. *shrug*

**Comment by josegonzalez** [comment]
So as a first stab at this - and maybe making our postgres situation worse - I created a beta "official" postgres plugin that mostly adheres to the spec. I used `service:destroy` instead of `service:delete` since that's what we do for apps :)

Not all commands are implemented yet, though I do plan to do so before I consider the plugin "stable". Also note that this plugin is *not* intended for wide use just yet, so ymmv and all that jazz. I'll add it to the plugin index once I think it's ready for primetime.

https://github.com/dokku/dokku-postgres-plugin
